core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -62,6 +65,39 @@ class LogCleanerTest {
     Utils.delete(tmpdir)
   }
 
+  @Test
+  def testRemoveMetricsOnClose(): Unit = {
+    val mockMetricsGroupCtor = mockConstruction(classOf[KafkaMetricsGroup])
+    try {
+      val logCleaner = new LogCleaner(new CleanerConfig(true),
+        logDirs = Array(TestUtils.tempDir()),
+        logs = new Pool[TopicPartition, UnifiedLog](),
+        logDirFailureChannel = new LogDirFailureChannel(1),
+        time = time)
+
+      // shutdown logCleaner so that metrics are removed
+      logCleaner.shutdown()
+
+      val mockMetricsGroup = mockMetricsGroupCtor.constructed.get(0)
+      val numMetricsRegistered = 5
+      verify(mockMetricsGroup, times(numMetricsRegistered)).newGauge(anyString(), any())
```

Comment:
It is intentionally left as generic so that this can capture all `gauge` in this class. If someone adds a gauge and doesn't update this test, this line will fail. Hence, catching all generic strings makes this test even more secure. 

We don't need to do something similar for remove because, we validate that the number of "added" is equal to the number of metrics names in the next statement. This ensures that all added metrics are in MetricNames and later we validate that all MetricNames are closed.

Reply:
That makes sense. Thanks for the clarification. I wonder if we could use `LogCleaner.MetricNames.size` instead of hardcoding `5`. I suppose that it would have a similar semantic, does it?

Reply:
Your suggestion will work (for now) until all metrics in this class of type `gauge` but later if someone introduces a metric of another type and adds it to MetricNames, it will fail.

Nevertheless, I have made the change as you suggested. The test could be changed later when someone adds a new metric.